pkg/kademlia/lookup.go

```diff
@@ -38,60 +44,116 @@ func newSequentialLookup(rt *RoutingTable, nodes []*pb.Node, client node.Client,
 		heap.Init(&queue)
 	}
 
-	return &sequentialLookup{
+	return &lookup{
+		lMutex:          &sync.Mutex{},
 		contacted:       map[string]bool{},
 		queue:           queue,
 		slowestResponse: 0,
 		client:          client,
 		target:          target,
-		limit:           limit,
-		bootstrap:       bootstrap,
+		routines:        defaultRoutines,
+		rMutex:          &sync.Mutex{},
 	}
 }
 
-func (lookup *sequentialLookup) Run(ctx context.Context) error {
-	zero := &big.Int{}
-	targetBytes := new(big.Int).SetBytes(lookup.target.Bytes())
-
-	for len(lookup.queue) > 0 {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-		}
+func (lookup *lookup) Run(ctx context.Context) error {
+	wg := sync.WaitGroup{}
+	nowork := sync.Cond{L: lookup.rMutex}
+	working := 0
+	allDone := false
+
+	wg.Add(lookup.routines)
+	for i := 0; i < lookup.routines; i++ {
+		go func(i int) {
+			defer func() {
+				wg.Done()
+			}()
+			for {
+				var item *Item
+
+				// protect the state with a lock
+				lookup.rMutex.Lock()
+				for {
+					// when everything is done then we can just leave
+					if allDone {
+						lookup.rMutex.Unlock()
+						return
+					}
+
+					// try to get work
+					var ok bool
+					item, ok = lookup.queue.Pop().(*Item)
+					if ok {
+						// mark ourselves as working
+						working++
+						break
+					}
+
+					// no work, let's wait until some other worker pushes new work
+					nowork.Wait()
+				}
+				lookup.rMutex.Unlock()
+
+				// do the work and find new work items
+				targetBytes := new(big.Int).SetBytes(lookup.target.Bytes())
+				additionalWork := lookup.work(ctx, item, targetBytes)
+
+				// protect the state
+				lookup.rMutex.Lock()
+
+				// add things to the queue
+				for _, work := range additionalWork {
+					lookup.queue.Push(work)
+					// wake up one worker
+					nowork.Signal()
```

Comment:
Since we are using a wide lock and broadcast below, there's no reason to signal here.